library/x509_crt.c

```diff
@@ -620,231 +850,584 @@ static int x509_get_ext_key_usage( unsigned char **p,
+    if( ( ret = mbedtls_asn1_get_tag( &p, end, &len,
+                                      MBEDTLS_ASN1_OCTET_STRING ) ) != 0 )
+        goto err;
 
-        if( ret != 0 )
-        {
-            /* No parser found, skip extension */
-            *p = end_ext_octet;
+    end_ext_octet = p + len;
+    if( end_ext_octet != end )
+    {
+        ret = MBEDTLS_ERR_ASN1_LENGTH_MISMATCH;
+        goto err;
+    }
 
+    /*
+     * Detect supported extensions
+     */
+    ret = mbedtls_oid_get_x509_ext_type( &extn_oid, &ext_type );
+    if( ret != 0 )
+    {
 #if !defined(MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION)
-            if( is_critical )
-            {
-                /* Data is marked as critical: fail */
-                return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS +
-                        MBEDTLS_ERR_ASN1_UNEXPECTED_TAG );
-            }
-#endif
-            continue;
+        if( is_critical )
+        {
+            /* Data is marked as critical: fail */
+            ret = MBEDTLS_ERR_ASN1_UNEXPECTED_TAG;
+            goto err;
         }
+#endif
+        return( 0 );
+    }
 
-        /* Forbid repeated extensions */
-        if( ( crt->ext_types & ext_type ) != 0 )
-            return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS );
-
-        crt->ext_types |= ext_type;
+    /* Forbid repeated extensions */
+    if( ( frame->ext_types & ext_type ) != 0 )
+        return( MBEDTLS_ERR_X509_INVALID_EXTENSIONS );
 
-        switch( ext_type )
-        {
+    frame->ext_types |= ext_type;
+    switch( ext_type )
+    {
         case MBEDTLS_X509_EXT_BASIC_CONSTRAINTS:
+        {
+            int ca_istrue;
+            int max_pathlen;
+
             /* Parse basic constraints */
-            if( ( ret = x509_get_basic_constraints( p, end_ext_octet,
-                    &crt->ca_istrue, &crt->max_pathlen ) ) != 0 )
-                return( ret );
+            ret = x509_get_basic_constraints( &p, end_ext_octet,
+                                              &ca_istrue,
+                                              &max_pathlen );
+            if( ret != 0 )
+                goto err;
+
+            frame->ca_istrue   = ca_istrue;
+            frame->max_pathlen = max_pathlen;
             break;
+        }
 
         case MBEDTLS_X509_EXT_KEY_USAGE:
             /* Parse key usage */
-            if( ( ret = x509_get_key_usage( p, end_ext_octet,
-                    &crt->key_usage ) ) != 0 )
-                return( ret );
+            ret = x509_get_key_usage( &p, end_ext_octet,
+                                      &frame->key_usage );
+            if( ret != 0 )
+                goto err;
+            break;
+
+        case MBEDTLS_X509_EXT_SUBJECT_ALT_NAME:
+            /* Copy reference to raw subject alt name data. */
+            frame->subject_alt_raw.p   = p;
+            frame->subject_alt_raw.len = end_ext_octet - p;
+
+            ret = mbedtls_asn1_traverse_sequence_of( &p, end_ext_octet,
```

Comment:
Yes, this function only checks well-formedness of the ASN.1 SEQUENCE, but doesn't do any further parsing. This is deliberate: We cannot parse what we don't know. 

When we start supporting structured SAN types, we have two options: (1) Change the callback here to ensure structural integrity of the SAN components we know (nested parsing). (2) Do the SAN component parsing afterwards.

I am inclined to (1), which I think would be in the spirit of this PR.

Reply:
No change (for now).

Reply:
changing the callback here will not do, because the callback will be called only if `tag & MBEDTLS_ASN1_TAG_VALUE_MASK == 2`
This means that once we introduce a new type (which will happen very soon), this function will need to change rather than a new callback.
This also means that the usage of `mbedtls_asn1_traverse_sequence_of` here should not be done, as it doesn't cover all possible cases. (even other well formed ASN.1 sequences such as IP address)

Reply:
I don't see the issue - yes, this code will need to be adapted soon when we introduce new SAN types, but for now, it maintains the previous behaviour.

Reply:
The issue is that we are introducing something that we know is not correct

Reply:
Apologies, Ron, but I don't follow and I cannot add to what I said before.